File: src/RotateArray.java

public class RotateArray {
	public void rotate(int[] nums, int k) throws Exception{
        int length=nums.length;
        k=k%length;
        revert(nums,0,length-1);
        revert(nums,0,k-1);
        revert(nums,k,length-1);
        
    }
    
    private void revert(int[]nums, int start, int end) throws Exception{
        if(end-start>nums.length||end<start|| start<0||end>=nums.length)
            throw new Exception("Invalid Boundary");
        while(start<end){
            int temp=nums[start];
            nums[start]=nums[end];
            nums[end]=temp;
            ++start;
            --end;
        }
    }
	public static void main(String[] args) throws Exception {
		int[]nums={1,2,3,4,5};
		new RotateArray().rotate(nums, 3);
		for(int i:nums){
			System.out.println(i+" ");
		}
	}

}

File: python/rotate_array.py
__author__ = 'kaijiezhou'


class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        k%=len(nums)
        self.revert_arr(nums,0, len(nums)-1)
        self.revert_arr(nums, 0, k-1)
        self.revert_arr(nums, k, len(nums)-1)

    def revert_arr(self, nums, start, end):
        while start < end:
            tmp = nums[start]
            nums[start] = nums[end]
            nums[end] = tmp
            start += 1
            end -= 1
